hack/generator/pkg/astmodel/storage_conversion_factories.go

```diff
@@ -0,0 +1,360 @@
+
+	if conversion == nil {
+		// No conversion between the elements of the array
+		return nil
+	}
+
+	return func(reader dst.Expr, writer dst.Expr, ctx *CodeGenerationContext) []dst.Stmt {
+		itemId := sourceEndpoint.CreateSingularLocal()
+		indexId := itemId + "Index"
+		tempId := sourceEndpoint.CreatePluralLocal("List")
+
+		declaration := astbuilder.SimpleAssignment(
+			dst.NewIdent(tempId),
+			token.DEFINE,
+			astbuilder.MakeList(dt.AsType(ctx), astbuilder.CallFunc("len", reader)))
+
+		body := conversion(
+			dst.NewIdent(itemId),
+			&dst.IndexExpr{
+				X:     dst.NewIdent(tempId),
+				Index: dst.NewIdent(indexId),
+			},
+			ctx,
+		)
+
+		assign := astbuilder.SimpleAssignment(writer, token.ASSIGN, dst.NewIdent(tempId))
+
+		loop := astbuilder.IterateOverListWithIndex(indexId, itemId, reader, body...)
+		loop.Decs.After = dst.EmptyLine
+
+		return []dst.Stmt{
+			declaration,
+			loop,
+			assign,
+		}
+	}
+}
+
+// assignMapFromMap will generate a code fragment to populate an array, assuming the
+// underlying types of the two arrays are compatible
+//
+// <map> := make(map[<key>]<type>)
+// for key, <item> := range <reader> {
+//     <map>[<key>] := <item>
+// }
+// <writer> = <map>
+//
+func assignMapFromMap(
+	sourceEndpoint *StorageConversionEndpoint,
+	destinationEndpoint *StorageConversionEndpoint) StorageTypeConversion {
+	st := AsMapType(sourceEndpoint.Type())
+	dt := AsMapType(destinationEndpoint.Type())
+
+	if st == nil || dt == nil {
+		// One or other type is not a map
+		return nil
+	}
+
+	if !st.key.Equals(dt.key) {
+		// Keys are different types
+		return nil
+	}
+
+	srcEp := sourceEndpoint.WithType(st.value)
+	dstEp := destinationEndpoint.WithType(dt.value)
+	conversion, _ := createTypeConversion(srcEp, dstEp)
+
+	if conversion == nil {
+		// No conversion between the elements of the map
+		return nil
+	}
+
+	return func(reader dst.Expr, writer dst.Expr, ctx *CodeGenerationContext) []dst.Stmt {
+		itemId := sourceEndpoint.CreateSingularLocal()
+		keyId := itemId + "Key"
+		tempId := sourceEndpoint.CreatePluralLocal("Map")
+
+		declaration := astbuilder.SimpleAssignment(
+			dst.NewIdent(tempId),
+			token.DEFINE,
+			astbuilder.MakeMap(dt.key.AsType(ctx), dt.value.AsType(ctx)))
+
+		body := conversion(
+			dst.NewIdent(itemId),
+			&dst.IndexExpr{
+				X:     dst.NewIdent(tempId),
+				Index: dst.NewIdent(keyId),
+			},
+			ctx,
+		)
+
+		assign := astbuilder.SimpleAssignment(writer, token.ASSIGN, dst.NewIdent(tempId))
+
+		loop := astbuilder.IterateOverMapWithValue(keyId, itemId, reader, body...)
+		loop.Decs.After = dst.EmptyLine
+
+		return []dst.Stmt{
+			declaration,
+			loop,
+			assign,
+		}
+	}
+}
+
+func zeroValue(p *PrimitiveType) string {
+	switch p {
+	case StringType:
+		return "\"\""
+	case IntType:
+		return "0"
+	case FloatType:
+		return "0"
+	case UInt32Type:
+		return "0"
+	case UInt64Type:
+		return "0"
+	case BoolType:
+		return "false"
+	}
+
+	return "##DOESNOTCOMPUTE##"
```

Comment:
This feels like it should just be a panic to me? Better that than generating invalid code it feels like.

Reply:
Agreed.